### ISSUE TITLE
 jpegdecompresssurface : picture decode failed : e00002d1

### ISSUE THREAD

**Issue by totoourworld** [martinrybak/SQLClient#40]
jpegdecompresssurface : picture decode failed : e00002d1
I can get exact same data as image data from sql server.
So i try to load image from NSData, below the code,

- (void)process:(NSArray*)results
{
    NSArray *table=[results objectAtIndex:0];
    NSDictionary *row=[table objectAtIndex:0];
    NSData *image_source=[row objectForKey:@"Company_Logo"];
    UIImage *image=[[UIImage alloc] initWithData:image_source];
    image=[UIImage imageWithData:image_source];
    [self.imageView setImage :image];
}
but the final step , it is showing like this, in here
 // [self.imageView setImage :image];
jpegdecompresssurface : picture decode failed : e00002d1
How can i solve this?
Regards.

**Comment by martinrybak** [comment]
Please try the `types` branch.

**Comment by totoourworld** [comment]
I used types branch and used the sample project for testing but same.
can you give me the source code for that?
I checked the 
NSData *image_source=[row objectForKey:@"Company_Logo"];
by log the value.
but same as data in sql.
How can i use this for mapping to uiimage type?
can you teach me?
i want your reply. thanks .
king regards,

**Comment by martinrybak** [comment]
It could be related to this:
`https://msdn.microsoft.com/en-us/library/ms177687(v=sql.90).aspx`

What version of SQL Server are you using?
What data type is the column with the image?
If possible, can you attach the image file here?

**Comment by totoourworld** [comment]
I am using the SQL EXPRESS 2016.
I used the Company_Logo(image,null) for saving/loading image.
 
Additionally, I am trying to load the image that saved by C# program i made.
In android , the image can be read by byte array and convert that to bitmap image, and it is working.
In C# , it is also working well.
 
but in iOS when i tried to do like that with above code , it is showing 
jpegdecompresssurface : picture decode failed : e00002d1

**Comment by totoourworld** [comment]
![1](https://cloud.githubusercontent.com/assets/20016355/22238708/975dcad0-e245-11e6-83d3-16052e2fd09d.png)

**Comment by totoourworld** [comment]
![3d_apple_logo_preview_featured](https://cloud.githubusercontent.com/assets/20016355/22238806/05f3ec72-e246-11e6-86a0-78cf9d2ba94c.jpg)

**Comment by totoourworld** [comment]
![visual_csharp_logo1-300x230](https://cloud.githubusercontent.com/assets/20016355/22238821/14bb2b12-e246-11e6-9430-147b5543a60e.png)

**Comment by totoourworld** [comment]
![images](https://cloud.githubusercontent.com/assets/20016355/22238825/1f6e6f42-e246-11e6-8945-252c2a6e18c2.jpg)

**Comment by totoourworld** [comment]
These three pictures , i used to try.
jpg file looks like that issue , but png file is not showing at all.

**Comment by totoourworld** [comment]
Hello , Mr.Martin Rybak.
I tried to debug the code and found something related.
So, (void)process:(NSArray*)results
{
NSArray *table=[results objectAtIndex:0];
NSDictionary *row=[table objectAtIndex:0];
NSData *image_source=[row objectForKey:@"Company_Logo"];
UIImage *image=[[UIImage alloc] initWithData:image_source];
image=[UIImage imageWithData:image_source];
[self.imageView setImage :image];
}
in this code ,(NSArray*) results is returning values from sql.
in here , i checked the @"Company_Logo".
It was NSInlineData and 4096bytes.
That's why the above 5kb jpg image is not showing full.
I tried to use small jpg image (for example 400bytes) , and it is showing well.
Can you give me some solution for large image?
Really thanks for your help.
Regards.

**Comment by martinrybak** [comment]
Please put a breakpoint on line 544 of `SQLClient.m`:

https://github.com/martinrybak/SQLClient/blob/types/SQLClient/SQLClient/SQLClient/SQLClient.m#L544

What is the value of `column->size`?

**Comment by totoourworld** [comment]
in image type the column size is 4096

**Comment by totoourworld** [comment]
column->size is 4096

**Comment by martinrybak** [comment]
OK, to be sure, put a breakpoint on this line and verify that `dbcollen(_connection, c)` is returning 4096 for your image column.

https://github.com/martinrybak/SQLClient/blob/types/SQLClient/SQLClient/SQLClient/SQLClient.m#L261

**Comment by martinrybak** [comment]
Also try changing your data type from `image` to `varbinary(max)`

**Comment by totoourworld** [comment]
I tried both type, means , i saved the data to image type and varbinary(max) type
and read the data using your framework.
In 261 , breakpoint, the column.size is 4096 for both types , sir.

**Comment by martinrybak** [comment]
OK, in the freetds docs I found the following:

`text size`: default value of TEXTSIZE, in bytes. For text and image datatypes, sets the maximum width of any returned column. Cf. set TEXTSIZE in the T-SQL documentation for your server.

http://www.freetds.org/userguide/freetdsconf.htm

Try adding a file called `freetds.conf` to your app with these contents:
```
[global]
text size = 10000
```

Then add an environment variable to your app called `FREETDS` whose value is the path to that file.

http://www.freetds.org/userguide/envvar.htm

See if that helps.

**Comment by totoourworld** [comment]
sorry but my app means?
ios project, sir?

**Comment by martinrybak** [comment]
Yes, add it to your iOS project.

**Comment by totoourworld** [comment]
yeah ,i added freetds.conf to SQLClint folder , but i am not sure about the environment variable adding , can you guide me a bit more , sir?
thanks.

**Comment by martinrybak** [comment]
http://nshipster.com/launch-arguments-and-environment-variables/

**Comment by totoourworld** [comment]
freetds.conf path would be full path like this "/Volumes/Mac_Data/KO/pos_australia/latest/SQLClient_Type/SQLClient-types/SQLClient/SQLClient/SQLClient/freetds.conf" 
or what's the right path of freetds.conf file, sir?

**Comment by totoourworld** [comment]
I added with your requirement and ran the project
ntext type in sql returning column->size 8192 bytes in  dbcollen(_connection, c) 164 line.
but image type and varbinary is same as 4096, sir.

**Comment by martinrybak** [comment]
http://stackoverflow.com/questions/12205785/getting-bundle-file-references-paths-at-app-launch

**Comment by totoourworld** [comment]
I got file path using your reference source code , 
"/var/containers/Bundle/Application/0E8B03A5-869A-4435-AE41-21972F55CA4B/SQLClient.app/freetds.conf"
so i set above path as value of FREETDS.

**Comment by totoourworld** [comment]
what value would i check by debugging?

**Comment by totoourworld** [comment]
![00b9eedc-325f-4313-95ba-c97b0a649ffa](https://cloud.githubusercontent.com/assets/20016355/22264822/f5ce0ea2-e2ab-11e6-82ad-51664c017ffa.png)
![9f1448e2-d6b8-4b26-acb7-4323ae25c50f](https://cloud.githubusercontent.com/assets/20016355/22264821/f5cc4c48-e2ab-11e6-8925-81b5845f62c2.png)
![14048687-816f-4288-bfd4-e90db90ac44d](https://cloud.githubusercontent.com/assets/20016355/22264823/f5f90ae4-e2ab-11e6-8dda-2fb8d2abb277.png)

**Comment by martinrybak** [comment]
The value must be relative to your app. Try `freetds.conf` instead.

**Comment by totoourworld** [comment]
I changed with your instruction
and tested , again . but the column->size of image is still 4096

**Comment by totoourworld** [comment]
![ef9826a7-0c22-496b-91c7-dd11a662395a](https://cloud.githubusercontent.com/assets/20016355/22265282/017f58da-e2ae-11e6-8394-26ee4ce8ca2c.png)
![4f0e0644-3328-4224-bb90-386516836e4c](https://cloud.githubusercontent.com/assets/20016355/22265283/03465fb0-e2ae-11e6-90fd-164ad331a77d.png)

**Comment by martinrybak** [comment]
Try setting other values to verify that TDS is finding your file.
http://www.freetds.org/userguide/freetdsconf.htm

**Comment by totoourworld** [comment]
i setup the sql express 2016 to other computer windows 10 system.
in here , TDS version how can i check this?

I am really sorry for bothering your time, but can you teach me via my screen? like using teamviewer,sir?

**Comment by martinrybak** [comment]
I'm sorry, I can't do that. Please do some reading up on FreeTDS until you can customize the `text size` setting. I think that is the problem.

**Comment by totoourworld** [comment]
so the conclsion is change the text size?

**Comment by amiparaharikant** [comment]
i facing this issue in my iOS app. please help me out. thank you.

in android application is working fine to get binary image from SQL server but facing issue in iOS and load some part of image only.

**Comment by totoourworld** [comment]
can you show me the code?

**Comment by amiparaharikant** [comment]
![screen shot 2018-08-04 at 11 38 59 am](https://user-images.githubusercontent.com/11538768/43673152-08f9a852-97db-11e8-8055-15ec8eb76139.png)
@totoourworld i'm getting size of image is still 4096

**Comment by martinrybak** [comment]
https://stackoverflow.com/questions/4500506/php-is-truncating-mssql-blob-data-4096b-even-after-setting-ini-values-am-i-m

**Comment by amiparaharikant** [comment]
is this issue of getting size of image 4096byte is from SQL SERVER ?
